### PR TITLE
Add documentation for the RichText component to the Block Editor Handbook

### PR DESCRIPTION
**Pull request by jg314**
## Description
Add RichText component documentation to the Block Editor Handbook. Currently all that exists is the Github documentation at https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/rich-text/README.md.

## How has this been tested?
I did not test this outside of the basic code example provided in the content. If there is other testing that's needed please let me know.

## Types of changes

1. Create a new RichtText Reference page that lives under the Developer Documentation section of the handbook.
2. Modify the table of contents to include the new page.

## Other Notes

- This content was recommended in https://github.com/WordPress/gutenberg/issues/14528 by @noisysocks. 
- I attempted to run `npm run docs:build` as noted on https://github.com/WordPress/gutenberg/blob/master/docs/contributors/document.md, but I continued to get an error within terminal. Let me know if it would be helpful for me to troubleshoot that.
- Most importantly, this is the first time I've ever contributed, so I apologize in advance if I missed something. Please let me know if there are ways to improve what's here. I'd love and welcome any feedback.

### CODE REVIEW

**Comment by noisysocks** [comment]
Hey @jg314! Thanks so much for working on this!

> I attempted to run npm run docs:build as noted on https://github.com/WordPress/gutenberg/blob/master/docs/contributors/document.md, but I continued to get an error within terminal. Let me know if it would be helpful for me to troubleshoot that.

You'll need to generate docs and commit the files that it generates so that Travis CI is happy with this branch and lets us merge it.

What error are you receiving when you run `npm run docs:build`?

File: docs/designers-developers/developers/richtext.md
Comment (noisysocks) [inline docs/designers-developers/developers/richtext.md]: Since this will be published on WordPress.org, we should link to the handbook page here instead of linking to GitHub.

```suggestion
Unlike other components that exist in the [Component Reference](https://developer.wordpress.org/block-editor/components/) section, RichText lives separately because it only makes sense within the block editor, and not within other areas of WordPress.
```

File: docs/designers-developers/developers/richtext.md
Comment (noisysocks) [inline docs/designers-developers/developers/richtext.md]: Let's use the full URL here, since a relative one won't work when this is on WordPress.org.

```suggestion
For a list of the possible properties to pass your RichText component, [check out the component documentation on Github](https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/rich-text/README.md).
```

File: docs/designers-developers/developers/richtext.md
Comment (noisysocks) [inline docs/designers-developers/developers/richtext.md]: Let's make these four URLs absolute URLs as well.

**Review by noisysocks — CHANGES_REQUESTED** [review]

File: docs/designers-developers/developers/richtext.md
Comment (gziolo) [reply docs/designers-developers/developers/richtext.md]: It’s fine as is. See our guidelines:
https://github.com/WordPress/gutenberg/blob/master/docs/contributors/document.md#using-links

We should never optimize for the handbook. It has to work on GitHub as well. There might be at least a few weeks delay with updates on WordPress.org when we are waiting for the next major release.

File: docs/designers-developers/developers/richtext.md
Comment (gziolo) [reply docs/designers-developers/developers/richtext.md]: Again, it’s perfectly fine as is.

File: docs/designers-developers/developers/richtext.md
Comment (gziolo) [reply docs/designers-developers/developers/richtext.md]: Nope, let’s keep them 😅

File: docs/designers-developers/developers/richtext.md
Comment (gziolo) [reply docs/designers-developers/developers/richtext.md]: Well, maybe not. It’s very unfortunate. Ideally, this should stay as is and the handbook should point to the branch with the current version of WordPress rather than to master which can have features that might even not be published in the olugin.

File: docs/designers-developers/developers/richtext.md
Comment (gziolo) [reply docs/designers-developers/developers/richtext.md]: Or maybe not 🙁 I don’t know anymore. See my other comments.

**Comment by jg314** [comment]
@noisysocks, thanks for taking such a quick look at this PR. I've pasted the error I'm getting when trying to run `npm run docs:build` below:

```
11 silly lifecycle gutenberg@5.8.0~docs:build: Returned: code: 1  signal: null
12 info lifecycle gutenberg@5.8.0~docs:build: Failed to exec docs:build script
13 verbose stack Error: gutenberg@5.8.0 docs:build: `node ./docs/tool && node ./bin/update-readmes`
13 verbose stack Exit status 1
13 verbose stack     at EventEmitter.<anonymous> (C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\index.js:285:16)
13 verbose stack     at emitTwo (events.js:126:13)
13 verbose stack     at EventEmitter.emit (events.js:214:7)
13 verbose stack     at ChildProcess.<anonymous> (C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\lib\spawn.js:55:14)
13 verbose stack     at emitTwo (events.js:126:13)
13 verbose stack     at ChildProcess.emit (events.js:214:7)
13 verbose stack     at maybeClose (internal/child_process.js:925:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:209:5)
14 verbose pkgid gutenberg@5.8.0
15 verbose cwd C:\Users\username_removed\Desktop\gutenberg
16 verbose Windows_NT 10.0.17763
17 verbose argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "run" "docs:build"
18 verbose node v8.11.3
19 verbose npm  v5.6.0
20 error code ELIFECYCLE
21 error errno 1
22 error gutenberg@5.8.0 docs:build: `node ./docs/tool && node ./bin/update-readmes`
22 error Exit status 1
23 error Failed at the gutenberg@5.8.0 docs:build script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 1, true ]
```

Have you all seen this before? Any idea what I can do to fix it?

File: docs/designers-developers/developers/richtext.md
Comment (jg314) [reply docs/designers-developers/developers/richtext.md]: Let me know which approach you all want me to use on this front. I had followed the guidelines on https://github.com/WordPress/gutenberg/blob/master/docs/contributors/document.md#using-links, but I'm happy to make these absolute URLs if that makes more sense.

File: docs/designers-developers/developers/richtext.md
Comment (gziolo) [reply docs/designers-developers/developers/richtext.md]: @nosolosw or @chrisvanpatten - any thought how we should approach it? I'm happy to apply the necessary changes to [Devhub](https://make.wordpress.org/docs/handbook/devhub/) and document it.

File: docs/designers-developers/developers/richtext.md
Comment (nosolosw) [reply docs/designers-developers/developers/richtext.md]: This link is fine.

If in doubt, to know whether a link will work in the handbook you can 1) remove the `README.md` part and 2) add it to `https://developer.wordpress.org/block-editor/` base URL. For example, in this case, navigate to https://developer.wordpress.org/block-editor//packages/components/ and notice how it's redirected to the proper URL.

File: docs/designers-developers/developers/richtext.md
Comment (nosolosw) [reply docs/designers-developers/developers/richtext.md]: This link won't work (can be checked with the suggestion above).

In general, links to subdirectories of top-level sections aren't exposed in the handbook. We should expose block-editor components in the handbook, and there is https://github.com/WordPress/gutenberg/issues/15107 to track that. I'm unsure what's the status of that.

For this kind of thing, I see that we're using a link directly to the `master` branch (searched for `master` within the docs folder), so I guess it's OK until we figure how to pin the handbook to a specific version.

File: docs/designers-developers/developers/richtext.md
Comment (nosolosw) [reply docs/designers-developers/developers/richtext.md]: Same as https://github.com/WordPress/gutenberg/pull/15956/files#r290299752 - These should be absolute URLs, as we don't expose the source code in the handbook.

**Comment by nosolosw** [comment]
This is a great addition! 

The RichText component API is exposed in the handbook as part of the block-editor package. See https://developer.wordpress.org/block-editor/packages/packages-block-editor/#RichText 

Unfortunately, it is undocumented and points to its README. This can be fixed by updating the [source JSDoc comment](https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/rich-text/index.js#L1263) and the changes will be reflected in the handbook.

That'd be a great separate PR that would play well with the good work on this one :)

File: docs/designers-developers/developers/richtext.md
Comment (jg314) [reply docs/designers-developers/developers/richtext.md]: Thanks for all the details. Just to confirm, it sounds like we should leave this one as a relative link without the suggested change. Is that correct?

File: docs/designers-developers/developers/richtext.md
Comment (jg314) [reply docs/designers-developers/developers/richtext.md]: For this one it sounds like we need to move forward using the absolute URL suggested by @noisysocks. Is that correct?

File: docs/designers-developers/developers/richtext.md
Comment (jg314) [reply docs/designers-developers/developers/richtext.md]: Great, it sounds like we should go ahead and make these absolute links. I can take care of that.

**Comment by jg314** [comment]
Thanks for the feedback @nosolosw. It sounds like once we've figured out the issue with `npm run docs:build` and solidified the plans for these relative vs. absolute URLs we should be good to move forward.

File: docs/designers-developers/developers/richtext.md
Comment (gziolo) [reply docs/designers-developers/developers/richtext.md]: Correct :)

File: docs/designers-developers/developers/richtext.md
Comment (gziolo) [reply docs/designers-developers/developers/richtext.md]: Yes, we should follow what @noisysocks shared.

**Comment by jg314** [comment]
The issue I mentioned above regarding `npm run docs:build` seems to be related to `execSync( join( __dirname, 'update-data.js' ) );` within wp-content\plugins\gutenberg\docs\tool\index.js. I'm on a Windows machine and spaces in directory names are causing the error. Any suggestions on how to deal with that? Thanks a ton for the help.

**Comment by noisysocks** [comment]
Hey @jg314, it's likely a bug in the docs tool. I'm setting up a Windows environment and looking into it now 🙂

**Comment by noisysocks** [comment]
Hi @jg314, I opened https://github.com/WordPress/gutenberg/pull/16029 which should fix the bug you're running into. Once my PR is merged you ought to be able to rebase your PR, generate docs, and get Travis CI to pass! 😅

File: docs/designers-developers/developers/richtext.md
Comment (jg314) [reply docs/designers-developers/developers/richtext.md]: These links have been updated to be absolute URLs. I'll go ahead and mark this as resolved.

**Comment by jg314** [comment]
The necessary URLs have been updated to be absolute. Once I'm able to run `npm run docs:build` without error we should be set to move this forward.

Thanks for all the help on this so far everyone.

**Comment by gziolo** [comment]
> The necessary URLs have been updated to be absolute.

All URLs works properly, I'm glad we managed to clarify what's the proper approach. It's still not ideal but at least we have an agreement :)

**Comment by jg314** [comment]
I spent some more time today trying to run `npm run docs:build` with no luck. I attempted to work through the issues noted here and in #16029, but I'm still getting errors when I run the command. Does anyone have any suggestions on how to move this forward? I'm absolutely open to a different approach. Thanks for the help.

**Comment by jg314** [comment]
I was finally able to get `npm run docs:build` to work by following these steps:

1. Moving the gutenberg plugin folder to another location (I did this because one of the directories where the code lives had a space in it, causing the build to fail)
2. Installing the Windows Subsystem for Linux
3. Running `npm install --save-dev react-docgen` (I did this because I was getting an error that said `.../node_modules/.bin/docgen: not found`)

Once it runs a lot of `.md` files are modified to remove documentation and replace it with "Undocumented declaration." That doesn't seem right, so I wanted to check before committing all of these files. Please let me know how you'd prefer I handle this. Thanks a ton.

**Comment by jg314** [comment]
Hey @noisysocks and @gziolo, is there anything else I can do to move this forward? Thanks again for all the help on this front.

**Comment by noisysocks** [comment]
Hey @jg314, sorry for the delay—I've been away on vacation!

> Once it runs a lot of .md files are modified to remove documentation and replace it with "Undocumented declaration." That doesn't seem right, so I wanted to check before committing all of these files.

No, that doesn't seem right 😕

This should be the result of `git diff` after you run `npm run docs:build`:

```diff
diff --git a/docs/manifest-devhub.json b/docs/manifest-devhub.json
index f762d1422..754744fcd 100644
--- a/docs/manifest-devhub.json
+++ b/docs/manifest-devhub.json
@@ -137,6 +137,12 @@
                "markdown_source": "../docs/designers-developers/developers/slotfills/plugin-sidebar-more-menu-item.md",
                "parent": "slotfills"
        },
+       {
+               "title": "RichText Reference",
+               "slug": "richtext",
+               "markdown_source": "../docs/designers-developers/developers/richtext.md",
+               "parent": "developers"
+       },
        {
                "title": "Internationalization",
                "slug": "internationalization",
```

Let's try to get everything fixed up over in https://github.com/WordPress/gutenberg/pull/16029 and then we can continue with this PR 🙂 — thanks for your patience!

**Comment by jg314** [comment]
Thanks for getting back to me @noisysocks. I really appreciate it. And that sounds good, let's work on #16029 first and then we'll get to this one.

**Comment by jg314** [comment]
After the awesome work @noisysocks did on https://github.com/WordPress/gutenberg/pull/16029 I was able to commit up the modified manifest.json file and the checks have passed. @noisysocks, is there anything else I can do to help at this point?

**Review by noisysocks — APPROVED** [review]

**Comment by noisysocks** [comment]
Thanks for your patience @jg314 and for contributing some documentation! This looks like a great start to get in and iterate on 🙂

**Comment by jg314** [comment]
Awesome, I'm happy to see it live. Thanks again for all the help @noisysocks. I really appreciate it.

**Pull request merged**
